bridge/src/helpers.rs

```diff
@@ -17,40 +17,39 @@
 //! various helper functions
 
 use error::{self, ResultExt};
-use ethabi::{self, ContractFunction, ParseLog, RawLog};
+use ethabi::{self, RawLog, FunctionOutputDecoder};
 use futures::future::FromErr;
 use futures::{Async, Future, Poll, Stream};
 use serde::de::Error;
 use serde::{Deserialize, Deserializer, Serializer};
 use std::time::Duration;
 use tokio_timer::{Timeout, Timer};
 use web3::api::Namespace;
-use web3::helpers::CallResult;
+use web3::helpers::CallFuture;
 use web3::types::{Address, Bytes, CallRequest, H256, TransactionRequest, U256};
 use web3::{self, Transport};
 
 /// attempts to convert a raw `web3_log` into the ethabi log type of a specific `event`
-pub fn parse_log<T: ParseLog>(event: &T, web3_log: &web3::types::Log) -> ethabi::Result<T::Log> {
+pub fn parse_log<T: Fn(RawLog) -> ethabi::Result<L>, L>(parse: T, web3_log: &web3::types::Log) -> ethabi::Result<L> {
     let ethabi_log = RawLog {
         topics: web3_log.topics.iter().map(|t| t.0.into()).collect(),
         data: web3_log.data.0.clone(),
     };
-    event.parse_log(ethabi_log)
+    parse(ethabi_log)
 }
 
 /// use `AsyncCall::new(transport, contract_address, timeout, function)` to
 /// get a `Future` that resolves with the decoded output from calling `function`
 /// on `contract_address`.
-pub struct AsyncCall<T: Transport, F: ContractFunction> {
-    future: Timeout<FromErr<CallResult<Bytes, T::Out>, error::Error>>,
+pub struct AsyncCall<T: Transport, F: FunctionOutputDecoder> {
+    future: Timeout<FromErr<CallFuture<Bytes, T::Out>, error::Error>>,
     function: F,
 }
 
-impl<T: Transport, F: ContractFunction> AsyncCall<T, F> {
+impl<T: Transport, F: FunctionOutputDecoder> AsyncCall<T, F> {
     /// call `function` at `contract_address`.
     /// returns a `Future` that resolves with the decoded output of `function`.
-    pub fn new(transport: &T, contract_address: Address, timeout: Duration, function: F) -> Self {
-        let payload = function.encoded();
+    pub fn new(transport: &T, contract_address: Address, timeout: Duration, payload: Vec<u8>, function: F) -> Self {
```

Comment:
`output_decoder` would be a more descriptive name for the last argument than `function`